src/System.Diagnostics.PerformanceCounter/src/System/Diagnostics/PerformanceData/CounterSet.cs

```diff
@@ -0,0 +1,296 @@
+                    throw new ArgumentException(SR.Format(SR.Perflib_Argument_CounterAlreadyExists, counterId, m_counterSet), nameof(counterId));
+                }
+
+                m_stringToId.Add(counterName, counterId);
+                m_idToCounter.Add(counterId, counterType);
+            }
+        }
+
+        /// <summary>
+        /// Create instances of the CounterSet. Created CounterSetInstance identifies active identity and tracks raw counter data for that identity.
+        /// </summary>
+        /// <param name="instanceName">Friendly name identifies the instance. InstanceName would be shown in counter consumption applications like perfmon.</param>
+        /// <returns>CounterSetInstance object</returns>
+        [SuppressMessage("Microsoft.Usage", "CA2208:InstantiateArgumentExceptionsCorrectly")]
+        public CounterSetInstance CreateCounterSetInstance(string instanceName)
+        {
+            if (instanceName == null)
+            {
+                throw new ArgumentNullException(nameof(instanceName));
+            }
+            if (instanceName.Length == 0)
+            {
+                throw new ArgumentException(SR.Format(SR.Perflib_Argument_EmptyInstanceName), nameof(instanceName));
+            }
+            if (m_provider == null)
+            {
+                throw new InvalidOperationException(SR.Format(SR.Perflib_InvalidOperation_NoActiveProvider, m_providerGuid));
+            }
+            if (!m_instanceCreated)
+            {
+                lock (m_lockObject)
+                {
+                    if (!m_instanceCreated)
+                    {
+                        if (m_provider == null)
+                        {
+                            throw new ArgumentException(SR.Format(SR.Perflib_Argument_ProviderNotFound, m_providerGuid), "ProviderGuid");
+                        }
+                        if (m_provider.m_hProvider.IsInvalid)
+                        {
+                            throw new InvalidOperationException(SR.Format(SR.Perflib_InvalidOperation_NoActiveProvider, m_providerGuid));
+                        }
+                        if (m_idToCounter.Count == 0)
+                        {
+                            throw new InvalidOperationException(SR.Format(SR.Perflib_InvalidOperation_CounterSetContainsNoCounter, m_counterSet));
+                        }
+
+                        uint Status = (uint)UnsafeNativeMethods.ERROR_SUCCESS;
+
+                        unsafe
+                        {
+                            uint CounterSetInfoSize = (uint)sizeof(UnsafeNativeMethods.PerfCounterSetInfoStruct)
+                                            + (uint)m_idToCounter.Count * (uint)sizeof(UnsafeNativeMethods.PerfCounterInfoStruct);
+                            uint CounterSetInfoUsed = 0;
+                            byte* CounterSetBuffer = stackalloc byte[(int)CounterSetInfoSize];
+
+                            if (CounterSetBuffer == null)
```

Comment:
I don't think this can happen. A stack overflow should happen if there is not enough space.

Reply:
fixed